presto-main/src/main/java/com/facebook/presto/sql/planner/PlanNodeCanonicalInfo.java

```diff
@@ -93,4 +94,24 @@ public static List<CanonicalPlanWithInfo> getCanonicalInfo(
         }
         return result.build();
     }
+
+    public static List<CanonicalPlanWithInfo> getPlanAnalyticsCanonicalInfo(
+            Session session,
+            PlanNode root,
+            PlanCanonicalInfoProvider planCanonicalInfoProvider)
+    {
+        ImmutableList.Builder<CanonicalPlanWithInfo> result = ImmutableList.builder();
+        for (PlanNode node : forTree(PlanNode::getSources).depthFirstPreOrder(root)) {
+            if (!node.getStatsEquivalentPlanNode().isPresent()) {
+                continue;
+            }
+            PlanNode statsEquivalentPlanNode = node.getStatsEquivalentPlanNode().get();
```

Comment:
I would not recommend using this field. It complicates the plan when things like  LimitNode are present(check it out in HistoricalStatisticsEquivalentPlanMarkingOptimizer).

Just compute hash of the final plan that you get querycompletedinfo. It will include exchanges, but you can take care of it. You can also skip caching - i don't get why we need it.

Reply:
@pranjalssh that makes sense, I will remove the caching.
 I think that there might be some more information that needs canonicalization in the Physical Plan other than exchanges. Also we need to add a new plan hasher with tests. Since the current one is production tested, I do not think there is a significant return on the ROI for now. Also in future we want to use this to detect  common CTE's and optimize their computation, so makes sense to still work on the logical Plan. I see that there is some complication with limit but I think we can still rely on the current HistoricalStatisticsEquivalentPlanMarkingOptimizer and fixes done in the current plan hasher to handle that case.

Reply:
@jaystarshot based on our production testing, size of canonical plan node increases a LOT when limit queries are present. And can even lead to timeouts when computing hashes for queries with big plans. For that case, we have added option to skip HBO for this type of query shapes. Hence, I'd recommend considering my sugeestion, and not have this overhead of simple query hashing. Hashing exchanges should be very easy.

If you still don't want to do it, I'm fine because its behind a flag that will be off by default. I suggest adding a nice comment for this. 

Also regarding caching, we don't need to cache plan statistics across queries. So create this cache for every query in getPlanAnalyticsQueryStats(that's the only place you need it)

Reply:
@pranjalssh I see thanks for explaining. But in the final physical plan there would be limits too, are you suggesting to avoid limit handling there?